### PR TITLE
VIITE-1772 New algorithm for finding the parallel link for the link w…

### PR DESCRIPTION
**Pull request by skosonen**
…ith discontinuity on two track road

### CODE REVIEW

File: digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/process/strategy/TrackCalculatorStrategy.scala [hunk 1]
```diff
@@ -211,22 +212,58 @@ trait TrackCalculatorStrategy {
...
+    *
+    * @param trackA
+    * @param linkOnTrackB
+    * @return (Project links before the discontinuity point, project links after the discontinuity point)
+    */
+  protected def getUntilNearestAddress(trackA: Seq[ProjectLink], linkOnTrackB: ProjectLink): (Seq[ProjectLink], Seq[ProjectLink]) = {
+    if (linkOnTrackB.discontinuity == MinorDiscontinuity || linkOnTrackB.discontinuity == Discontinuous) {
+
+      // Sort links in order by the distance from the end of the link on track b.
+      val linkOnTrackBEndPoint = endPointOfProjectLink(linkOnTrackB)
```
Comment (Campanha) [inline digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/process/strategy/TrackCalculatorStrategy.scala]: there is already defined methods for the first and last points for e.g.:
linkOnTrackB.getLastPoint
or
linkOnTrackB.getEndPoints()._2 (this one also checks if the link is reversed)

Comment (skosonen) [reply digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/process/strategy/TrackCalculatorStrategy.scala]: Thanks, I'll change it!

Comment (skosonen) [reply digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/process/strategy/TrackCalculatorStrategy.scala]: Fixed.

File: digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/process/strategy/TrackCalculatorStrategy.scala [hunk 2]
```diff
@@ -205,28 +206,58 @@ trait TrackCalculatorStrategy {
...
+
+      // Sort links in order by the distance from the end of the link on track b.
+      val linkOnTrackBEndPoint = linkOnTrackB.getEndPoints()._2
+      val nearestLinks: Seq[(ProjectLink, Double)] = trackA.map(pl => (pl,
+        pl.getEndPoints()._2.distance2DTo(linkOnTrackBEndPoint)
+      )).sortWith(_._2 < _._2) // Links with nearest end points first
+
+      // Choose the nearest link that has similar enough angle to the linkOnTrackB and set it as lastLink
+      val linkOnTrackBDirection = lastSegmentDirection(linkOnTrackB)
+      val maxAngleBetweenLinks = math.toRadians(45.0)
```
Comment (skosonen) [inline digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/process/strategy/TrackCalculatorStrategy.scala]: This 45 degrees might not be a correct value here. I'll continue testing and finding better value here.

File: digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/process/strategy/TrackCalculatorStrategy.scala [hunk 3]
```diff
@@ -205,28 +206,58 @@ trait TrackCalculatorStrategy {
...
+                                    startCP: Boolean, endCP: Boolean, source: CalibrationPointSource = UnknownSource): ProjectLink = {
     val sCP = if (startCP) CalibrationPointsUtils.makeStartCP(pl) else None
     val eCP = if (endCP) CalibrationPointsUtils.makeEndCP(pl, userCalibrationPoint) else None
     pl.copy(calibrationPoints = CalibrationPointsUtils.toProjectLinkCalibrationPointsWithSourceInfo((sCP, eCP), source))
   }
 
-  protected def getUntilNearestAddress(seq: Seq[ProjectLink], endProjectLink: ProjectLink): (Seq[ProjectLink], Seq[ProjectLink]) = {
-    if (endProjectLink.discontinuity == MinorDiscontinuity || endProjectLink.discontinuity == Discontinuous) {
-      val continuousProjectLinks = seq.takeWhile(pl => pl.startAddrMValue < endProjectLink.endAddrMValue)
+  private def lastSegmentDirection(pl: ProjectLink): Vector3d = {
```
Comment (skosonen) [inline digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/process/strategy/TrackCalculatorStrategy.scala]: This function is not working now correctly.

File: digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/process/strategy/TrackCalculatorStrategy.scala [hunk 4]
```diff
@@ -205,28 +206,50 @@ trait TrackCalculatorStrategy {
...
+
+      // Sort links in order by the distance from the end of the link on track b.
+      val linkOnTrackBEndPoint = linkOnTrackB.lastPoint
+      val nearestLinks: Seq[(ProjectLink, Double)] = trackA.map(pl => (pl,
+        pl.lastPoint.distance2DTo(linkOnTrackBEndPoint)
+      )).sortWith(_._2 < _._2) // Links with nearest end points first
+
+      // Choose the nearest link that has similar enough angle to the linkOnTrackB and set it as lastLink
+      val linkOnTrackBDirection = linkOnTrackB.lastSegmentDirection
+      val maxAngleBetweenLinks = math.toRadians(67.5) // TODO
```
Comment (pedrosag) [inline digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/process/strategy/TrackCalculatorStrategy.scala]: Please do address the magic number 67.5 and the //TODO

File: digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/process/strategy/TrackCalculatorStrategy.scala [hunk 5]
```diff
@@ -205,28 +206,50 @@ trait TrackCalculatorStrategy {
...
+
+      // Return links before the discontinuity point and links after it
+      val continuousLinks = trackA.takeWhile(pl => pl.endAddrMValue <= lastLink.endAddrMValue)
+      if (continuousLinks.isEmpty)
         throw new RoadAddressException("Could not find any nearest road address")
 
-      val lastProjectLink = continuousProjectLinks.last
-      if (continuousProjectLinks.size > 1 && lastProjectLink.toMeters(Math.abs(endProjectLink.endAddrMValue - lastProjectLink.startAddrMValue)) < lastProjectLink.toMeters(Math.abs(endProjectLink.endAddrMValue - lastProjectLink.endAddrMValue))) {
-        (continuousProjectLinks.init, lastProjectLink +: seq.drop(continuousProjectLinks.size))
+      // TODO
```
Comment (pedrosag) [inline digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/process/strategy/TrackCalculatorStrategy.scala]: What is left //TODO here?

File: digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/process/strategy/TrackCalculatorStrategy.scala (hunk 4 above)
Comment (skosonen) [reply digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/process/strategy/TrackCalculatorStrategy.scala]: It's now 75 degrees and there is a comment trying to explain it a bit.

**Comment by skosonen** [comment]
Tested locally on "triangle" and "rectangle" cases. Also tested the "banana" case on "rectangle" case. Seems to work fine in these cases. Tests pass on Jenkins (#266).

**Pull request merged**